planner/core/rule_join_reorder_greedy.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"math"
+	"sort"
+
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+)
+
+// extractJoinGroup will extract the nodes that connected by join tree.
+// A join tree is a tree that all node is join except the leaves.
+func extractJoinGroup(p LogicalPlan) (group []LogicalPlan, eqEdges []*expression.ScalarFunction, otherConds []expression.Expression) {
+	join, isJoin := p.(*LogicalJoin)
+	if !isJoin || join.preferJoinType > uint(0) || join.JoinType != InnerJoin || join.StraightJoin {
+		return []LogicalPlan{p}, nil, nil
+	}
+
+	lhsGroup, lhsEqualConds, lhsOtherConds := extractJoinGroup(join.children[0])
+	rhsGroup, rhsEqualConds, rhsOtherConds := extractJoinGroup(join.children[1])
+
+	group = append(group, lhsGroup...)
+	group = append(group, rhsGroup...)
+	eqEdges = append(eqEdges, join.EqualConditions...)
+	eqEdges = append(eqEdges, lhsEqualConds...)
+	eqEdges = append(eqEdges, rhsEqualConds...)
+	otherConds = append(otherConds, join.OtherConditions...)
+	otherConds = append(otherConds, lhsOtherConds...)
+	otherConds = append(otherConds, rhsOtherConds...)
+	return group, eqEdges, otherConds
+}
+
+type joinReOrderGreedySolver struct {
+	ctx          sessionctx.Context
+	curJoinGroup []LogicalPlan
+	eqEdges      []*expression.ScalarFunction
+	otherConds   []expression.Expression
+}
+
+// solve will construct a join tree by given join group using greedy algorithm.
+// We first choose the node with least cost. Then always find the node which can join with current tree and have least cost.
+func (s *joinReOrderGreedySolver) solve() (LogicalPlan, error) {
+	for _, node := range s.curJoinGroup {
+		_, err := node.deriveStats()
+		if err != nil {
+			return nil, err
+		}
+	}
+	sort.Slice(s.curJoinGroup, func(i, j int) bool {
```

Comment:
seems this sort is unnecessary?

Reply:
If the graph is totally connected, we don't need it.
But if the graph is not connected, just a collection of connected subgraph. Then for each subgraph, we need to find its smallest node.
With this sort, we just use `curJoinGroup[0]` is enough.
Without it, we need to always find the smallest one when we start a new subgraph.

Reply:
Got it.